File: sorting/bucket_sort.go
/*
	Bucket sort, or bin sort, is a sorting algorithm that works by distributing
	the elements of an array into a number of buckets. Each bucket is then
	sorted individually, either using a different sorting algorithm,
	or by recursively applying the bucket sorting algorithm.
	It is a distribution sort, a generalization of pigeonhole sort that
	allows multiple keys per bucket, and is a cousin of
	radix sort in the most-to-least significant digit flavor.
	Bucket sort can be implemented with comparisons and therefore can also
	be considered a comparison sort algorithm.
	The computational complexity depends on the algorithm used to sort
	each bucket, the number of buckets to use,
	and whether the input is uniformly distributed.

	Bucket sort works as follows:

	Set up an array of initially empty "buckets".
	Scatter: Go over the original array, putting each object in its bucket.
	Sort each non-empty bucket.
	Gather: Visit the buckets in order and put all elements back into the original array.

	Source(https://en.wikipedia.org/wiki/Bucket_sort)
*/
// The average time complexity for Bucket Sort is O(n + k).
// The worst time complexity is O(n²).
// The space complexity for Bucket Sort is O(n+k).
package main

import (
	"fmt"
)

func InsertionSort(Array []int) {
	for i := 0; i < len(Array); i++ {
		temp := Array[i]
		j := i - 1
		for ; j >= 0 && Array[j] > temp; j-- {
			Array[j+1] = Array[j]
		}
		Array[j+1] = temp
	}
}

func BucketSort(Array []int, bucketSize int) []int {
	var max, min int
	for _, n := range Array {
		if n < min {
			min = n
		}
		if n > max {
			max = n
		}
	}
	nBuckets := int(max-min)/bucketSize + 1
	buckets := make([][]int, nBuckets)
	for i := 0; i < nBuckets; i++ {
		buckets[i] = make([]int, 0)
	}

	for _, n := range Array {
		idx := int(n-min) / bucketSize
		buckets[idx] = append(buckets[idx], n)
	}

	sorted := make([]int, 0)
	for _, bucket := range buckets {
		if len(bucket) > 0 {
			InsertionSort(bucket)
			sorted = append(sorted, bucket...)
		}
	}

	return sorted
}

func main() {
	Array := []int{3, 4, 5, 2, 1}
	Array = BucketSort(Array, 2)
	for _, val := range Array {
		fmt.Println(val)
	}
}

File: sorting/bucket_sort.js
Here's an implementation of bucket sort in JavaScript:

function bucketSort(arr, numBuckets = 10) {
  const n = arr.length;
  const buckets = new Array(numBuckets);
  
  // Initialize the buckets
  for (let i = 0; i < numBuckets; i++) {
    buckets[i] = [];
  }
  
  // Distribute the elements into the buckets
  for (let i = 0; i < n; i++) {
    const bucketIndex = Math.floor(arr[i] * numBuckets);
    buckets[bucketIndex].push(arr[i]);
  }
  
  // Sort the elements within each bucket
  for (let i = 0; i < numBuckets; i++) {
    buckets[i].sort((a, b) => a - b);
  }
  
  // Concatenate the sorted buckets
  let sortedArr = [];
  for (let i = 0; i < numBuckets; i++) {
    sortedArr = sortedArr.concat(buckets[i]);
  }
  
  return sortedArr;
}


The bucketSort function takes an array arr of numbers and an optional parameter numBuckets that specifies the number of buckets to use (default is 10).
The algorithm works by distributing the elements of arr into numBuckets buckets based on their value. Each bucket contains elements within a range of values,
and since the buckets are sorted, the elements within each bucket are sorted as well. Finally, the sorted buckets are concatenated to obtain the final sorted
array.

Note that the performance of bucket sort depends on the distribution of the elements in the input array. If the elements are evenly distributed, bucket sort can achieve a time complexity of O(n), which is faster than many other sorting algorithms. However, if the elements are concentrated in a small range of values, the buckets may become unbalanced and the performance may degrade. Therefore, bucket sort is often used as a sub-routine in other sorting algorithms, or when the distribution of the input is known to be relatively even.
